File: Shared/RelaxKonOS.Protocol/ServerCenter/ServerReleaseValidation.cs
using System.Security.Cryptography;

namespace RelaxKonOS.Protocol.ServerCenter;

/// <summary>v1 发布签名算法。RSA-PSS + SHA-256 在 .NET 与 Android 均可用，因此作为必需算法。</summary>
public static class ServerReleaseSignatureAlgorithms
{
    public const string RsaPssSha256 = "rsa-pss-sha256";
}

/// <summary>
/// 发布清单、描述符与包布局的校验规则。客户端与启动器共用它，确保“包来源已确认”不是各端各写一套判断。
/// <para>签名语义：签名对象是制品的**原始文件字节**（<c>manifest.json</c> 或 <c>latest/{rid}.json</c>），
/// 签名单独放在同名 <c>*.sig</c> 文件中。<c>SignedSha256</c> 恒为这些原始字节的 SHA-256。
/// 这样 C#、Kotlin 与打包脚本不需要共享 JSON 规范化实现，避免跨语言字节不一致。</para>
/// </summary>
public static class ServerReleaseValidation
{
    /// <summary>校验包内清单结构。返回 null 表示通过，否则返回稳定的问题码。</summary>
    public static string? ValidateManifest(ServerReleaseManifestDto? manifest, ServerRuntimeIdentifier expectedRuntime)
    {
        if (manifest is null) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (manifest.SchemaVersion != ServerDeploymentProtocol.Version) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (!ServerDeploymentInputRules.IsVersion(manifest.Version)) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (manifest.Runtime != expectedRuntime) return ServerDeploymentProblemCodes.PackageRuntimeMismatch;
        if (manifest.SupportedSystems is null || manifest.SupportedSystems.Count == 0)
            return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (manifest.Files is null || manifest.Files.Count == 0 || manifest.Payload is null)
            return ServerDeploymentProblemCodes.PackageManifestInvalid;

        var platform = expectedRuntime is ServerRuntimeIdentifier.WinX64 or ServerRuntimeIdentifier.WinArm64
            ? "windows" : "linux";
        if (!manifest.Payload.TryGetValue(platform, out var payload) || payload.Count == 0)
            return ServerDeploymentProblemCodes.PackageManifestInvalid;

        var paths = new HashSet<string>(StringComparer.Ordinal);
        foreach (var file in manifest.Files)
        {
            var problem = ValidateFile(file);
            if (problem is not null) return problem;
            if (!paths.Add(file.Path)) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        }
        foreach (var path in payload.Values)
            if (!IsSafeManifestPath(path) || !paths.Contains(path))
                return ServerDeploymentProblemCodes.PackageManifestInvalid;
        return null;
    }

    /// <summary>校验清单结构并验证其伴随签名。</summary>
    public static string? VerifyManifest(
        ReadOnlySpan<byte> manifestBytes,
        ServerReleaseManifestDto? manifest,
        ServerRuntimeIdentifier expectedRuntime,
        ServerReleaseSignatureDto? signature,
        string publicKeyPem,
        ServerReleaseTrustPolicy trustPolicy)
    {
        var structural = ValidateManifest(manifest, expectedRuntime);
        if (structural is not null) return structural;
        return ServerReleaseSignatureVerifier.Verify(manifestBytes, signature, publicKeyPem, trustPolicy);
    }

    public static string? ValidateFile(ServerReleaseFileDto? file)
    {
        if (file is null) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (!IsSafeManifestPath(file.Path)) return ServerDeploymentProblemCodes.PackageLayoutUnsafe;
        if (!ServerDeploymentInputRules.IsSha256(file.Sha256)) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (file.Length < 0) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        return null;
    }

    /// <summary>
    /// 清单内路径必须是相对 POSIX 路径：无前导分隔符、无反斜杠、无 <c>..</c>、无空段。
    /// 解包时任何一条不满足都必须拒绝整个包，而不是跳过该条目。
    /// </summary>
    public static bool IsSafeManifestPath(string? path)
    {
        if (string.IsNullOrWhiteSpace(path)) return false;
        if (path.StartsWith('/') || path.StartsWith('\\')) return false;
        if (path.Contains('\\')) return false;
        if (path.Contains("..", StringComparison.Ordinal)) return false;
        if (path.Contains('\0')) return false;
        foreach (var segment in path.Split('/'))
            if (segment.Length == 0 || segment == ".") return false;
        return true;
    }

    /// <summary>校验下载描述符结构。签名是伴随文件，因此这里只做结构校验。</summary>
    public static string? ValidateDescriptor(ServerReleaseDescriptorDto? descriptor)
    {
        if (descriptor is null) return ServerDeploymentProblemCodes.PackageUnavailable;
        if (descriptor.SchemaVersion != ServerDeploymentProtocol.Version) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (!ServerDeploymentInputRules.IsVersion(descriptor.Version)) return ServerDeploymentProblemCodes.PackageManifestInvalid;
        if (!ServerDeploymentInputRules.IsSha256(descriptor.Sha256)) return ServerDeploymentProblemCodes.PackageDigestMismatch;
        return null;
    }

    /// <summary>校验描述符结构并验证其伴随签名。缺少签名直接拒绝。</summary>
    public static string? VerifyDescriptor(
        ReadOnlySpan<byte> descriptorBytes,
        ServerReleaseDescriptorDto? descriptor,
        ServerReleaseSignatureDto? signature,
        string publicKeyPem,
        ServerReleaseTrustPolicy trustPolicy)
    {
        var structural = ValidateDescriptor(descriptor);
        if (structural is not null) return structural;
        return ServerReleaseSignatureVerifier.Verify(descriptorBytes, signature, publicKeyPem, trustPolicy);
    }
}

/// <summary>
/// 发布签名校验。签名覆盖制品原始字节的 SHA-256，客户端用内置可信公钥验证，并确认 <c>keyId</c> 在信任策略内。
/// </summary>
public static class ServerReleaseSignatureVerifier
{
    /// <summary>
    /// 验证一段制品字节的签名。
    /// </summary>
    /// <param name="signedBytes">制品原始文件字节（<c>manifest.json</c> 或目录项 JSON）。</param>
    /// <param name="signature">签名记录，携带算法、keyId 与 Base64 签名。</param>
    /// <param name="publicKeyPem">该 keyId 对应的 SubjectPublicKeyInfo PEM。</param>
    /// <param name="trustPolicy">发布信任策略；未受信任的 keyId 直接拒绝。</param>
    public static string? Verify(
        ReadOnlySpan<byte> signedBytes,
        ServerReleaseSignatureDto? signature,
        string publicKeyPem,
        ServerReleaseTrustPolicy trustPolicy)
    {
        if (signature is null) return ServerDeploymentProblemCodes.PackageSignatureInvalid;
        if (signature.SchemaVersion != ServerDeploymentProtocol.Version) return ServerDeploymentProblemCodes.PackageSignatureInvalid;
        if (signature.Algorithm != ServerReleaseSignatureAlgorithms.RsaPssSha256)
            return ServerDeploymentProblemCodes.PackageSignatureInvalid;
        if (string.IsNullOrWhiteSpace(signature.KeyId) || string.IsNullOrWhiteSpace(signature.Signature))
            return ServerDeploymentProblemCodes.PackageSignatureInvalid;
        if (string.IsNullOrWhiteSpace(publicKeyPem)) return ServerDeploymentProblemCodes.PackageTrustRootMissing;
        if (!trustPolicy.IsTrustedKey(signature.KeyId) && !trustPolicy.AcceptsUntrustedSource)
            return ServerDeploymentProblemCodes.PackageSignatureInvalid;

        var digest = Convert.ToHexString(SHA256.HashData(signedBytes)).ToLowerInvariant();
        if (!string.Equals(digest, signature.SignedSha256, StringComparison.OrdinalIgnoreCase))
            return ServerDeploymentProblemCodes.PackageSignatureInvalid;

        byte[] signatureBytes;
        try { signatureBytes = Convert.FromBase64String(signature.Signature); }
        catch (FormatException) { return ServerDeploymentProblemCodes.PackageSignatureInvalid; }

        try
        {
            using var rsa = RSA.Create();
            rsa.ImportFromPem(publicKeyPem);
            var ok = rsa.VerifyData(signedBytes, signatureBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
            return ok ? null : ServerDeploymentProblemCodes.PackageSignatureInvalid;
        }
        catch (CryptographicException)
        {
            return ServerDeploymentProblemCodes.PackageSignatureInvalid;
        }
        catch (ArgumentException)
        {
            return ServerDeploymentProblemCodes.PackageSignatureInvalid;
        }
    }
}

File: Client/RelaxKonOS.Client.Android/app/src/main/java/app/relaxkonos/mobile/servercenter/ServerReleaseValidation.kt
package app.relaxkonos.mobile.servercenter

import java.security.KeyFactory
import java.security.MessageDigest
import java.security.PublicKey
import java.security.Signature
import java.security.spec.MGF1ParameterSpec
import java.security.spec.PSSParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.util.Locale

/** v1 发布签名算法。与 C# `ServerReleaseSignatureAlgorithms` 一致。 */
object ServerReleaseSignatureAlgorithms {
    const val RSA_PSS_SHA256 = "rsa-pss-sha256"
}

/**
 * 发布清单、描述符与包布局的校验规则。与 C# `ServerReleaseValidation` 对齐。
 *
 * 签名对象是制品的**原始文件字节**（`manifest.json` 或 `latest/{rid}.json`），签名单独放在同名 `*.sig` 文件中。
 * `signedSha256` 恒为这些原始字节的 SHA-256，因此 C#、Kotlin 与打包脚本不需要共享 JSON 规范化实现。
 */
object ServerReleaseValidation {

    /** 校验包内清单结构。返回 null 表示通过，否则返回稳定的问题码。 */
    fun validateManifest(manifest: ServerReleaseManifest?, expectedRuntime: ServerRuntimeIdentifier): String? {
        if (manifest == null) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (manifest.schemaVersion != ServerDeploymentProtocol.VERSION) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (!ServerDeploymentInputRules.isVersion(manifest.version)) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (manifest.runtime != expectedRuntime) return ServerDeploymentProblemCodes.PACKAGE_RUNTIME_MISMATCH
        if (manifest.supportedSystems.isEmpty()) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (manifest.files.isEmpty()) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        val platform = when (expectedRuntime) {
            ServerRuntimeIdentifier.WinX64, ServerRuntimeIdentifier.WinArm64 -> "windows"
            ServerRuntimeIdentifier.LinuxX64, ServerRuntimeIdentifier.LinuxArm64 -> "linux"
        }
        val payload = manifest.payload[platform]
            ?: return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (payload.isEmpty()) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        val paths = mutableSetOf<String>()
        for (file in manifest.files) {
            val problem = validateFile(file)
            if (problem != null) return problem
            if (!paths.add(file.path)) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        }
        if (payload.values.any { !isSafeManifestPath(it) || it !in paths }) {
            return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        }
        return null
    }

    /** 校验清单结构并验证其伴随签名。 */
    fun verifyManifest(
        manifestBytes: ByteArray,
        manifest: ServerReleaseManifest?,
        expectedRuntime: ServerRuntimeIdentifier,
        signature: ServerReleaseSignature?,
        publicKeyPem: String,
        trustPolicy: ServerReleaseTrustPolicy,
    ): String? {
        val structural = validateManifest(manifest, expectedRuntime)
        if (structural != null) return structural
        return ServerReleaseSignatureVerifier.verify(manifestBytes, signature, publicKeyPem, trustPolicy)
    }

    fun validateFile(file: ServerReleaseFile?): String? {
        if (file == null) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (!isSafeManifestPath(file.path)) return ServerDeploymentProblemCodes.PACKAGE_LAYOUT_UNSAFE
        if (!ServerDeploymentInputRules.isSha256(file.sha256)) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (file.length < 0) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        return null
    }

    /**
     * 清单内路径必须是相对 POSIX 路径：无前导分隔符、无反斜杠、无 `..`、无空段。
     * 解包时任何一条不满足都必须拒绝整个包。
     */
    fun isSafeManifestPath(path: String?): Boolean {
        if (path.isNullOrBlank()) return false
        if (path.startsWith('/') || path.startsWith('\\')) return false
        if (path.contains('\\')) return false
        if (path.contains("..")) return false
        if (path.contains('\u0000')) return false
        return path.split('/').none { it.isEmpty() || it == "." }
    }

    /** 校验下载描述符结构。签名是伴随文件，因此这里只做结构校验。 */
    fun validateDescriptor(descriptor: ServerReleaseDescriptor?): String? {
        if (descriptor == null) return ServerDeploymentProblemCodes.PACKAGE_UNAVAILABLE
        if (descriptor.schemaVersion != ServerDeploymentProtocol.VERSION) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (!ServerDeploymentInputRules.isVersion(descriptor.version)) return ServerDeploymentProblemCodes.PACKAGE_MANIFEST_INVALID
        if (!ServerDeploymentInputRules.isSha256(descriptor.sha256)) return ServerDeploymentProblemCodes.PACKAGE_DIGEST_MISMATCH
        return null
    }

    /** 校验描述符结构并验证其伴随签名。缺少签名直接拒绝。 */
    fun verifyDescriptor(
        descriptorBytes: ByteArray,
        descriptor: ServerReleaseDescriptor?,
        signature: ServerReleaseSignature?,
        publicKeyPem: String,
        trustPolicy: ServerReleaseTrustPolicy,
    ): String? {
        val structural = validateDescriptor(descriptor)
        if (structural != null) return structural
        return ServerReleaseSignatureVerifier.verify(descriptorBytes, signature, publicKeyPem, trustPolicy)
    }
}

/**
 * 发布签名校验。签名覆盖制品原始字节的 SHA-256，客户端用内置可信公钥验证，
 * 并确认 `keyId` 在信任策略内。
 */
object ServerReleaseSignatureVerifier {

    /**
     * 验证一段制品字节的签名。
     *
     * @param signedBytes 制品原始文件字节。
     * @param signature 签名记录，携带算法、keyId 与 Base64 签名。
     * @param publicKeyPem 该 keyId 对应的 SubjectPublicKeyInfo PEM。
     * @param trustPolicy 发布信任策略；未受信任的 keyId 直接拒绝。
     * @return null 表示通过，否则返回稳定的问题码。
     */
    fun verify(
        signedBytes: ByteArray,
        signature: ServerReleaseSignature?,
        publicKeyPem: String,
        trustPolicy: ServerReleaseTrustPolicy,
    ): String? {
        if (signature == null) return ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        if (signature.schemaVersion != ServerDeploymentProtocol.VERSION) return ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        if (signature.algorithm != ServerReleaseSignatureAlgorithms.RSA_PSS_SHA256) {
            return ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        }
        if (publicKeyPem.isBlank()) return ServerDeploymentProblemCodes.PACKAGE_TRUST_ROOT_MISSING
        if (!trustPolicy.isTrustedKey(signature.keyId) && !trustPolicy.acceptsUntrustedSource) {
            return ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        }

        val digest = sha256Hex(signedBytes)
        if (!digest.equals(signature.signedSha256, ignoreCase = true)) {
            return ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        }

        val signatureBytes = try {
            Base64Codec.decode(signature.signature)
        } catch (e: IllegalArgumentException) {
            return ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        }

        return try {
            val key = parsePublicKey(publicKeyPem)
            val verifier = Signature.getInstance("RSASSA-PSS")
            verifier.setParameter(PSSParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec.SHA256, 32, 1))
            verifier.initVerify(key)
            verifier.update(signedBytes)
            if (verifier.verify(signatureBytes)) null else ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        } catch (e: Exception) {
            ServerDeploymentProblemCodes.PACKAGE_SIGNATURE_INVALID
        }
    }

    private fun sha256Hex(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(bytes)
            .joinToString("") { "%02x".format(it) }

    private fun parsePublicKey(pem: String): PublicKey {
        val base64 = pem
            .replace("-----BEGIN PUBLIC KEY-----", "")
            .replace("-----END PUBLIC KEY-----", "")
            .filterNot { it.isWhitespace() }
        val der = Base64Codec.decode(base64)
        return KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(der))
    }
}

/**
 * 最小 Base64 编解码器（标准字母表，容忍缺失填充）。使用自有实现是为了同时满足：
 * `java.util.Base64` 需要 API 26，而 `android.util.Base64` 在 JVM 单元测试中不可用。
 */
internal object Base64Codec {
    private const val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

    fun decode(value: String): ByteArray {
        val filtered = value.filterNot { it.isWhitespace() }.trimEnd('=')
        if (filtered.isEmpty()) return ByteArray(0)
        val output = java.io.ByteArrayOutputStream()
        var buffer = 0
        var bits = 0
        for (c in filtered) {
            val index = ALPHABET.indexOf(c)
            require(index >= 0) { "Invalid base64 character: $c" }
            buffer = (buffer shl 6) or index
            bits += 6
            if (bits >= 8) {
                bits -= 8
                output.write((buffer shr bits) and 0xFF)
            }
        }
        return output.toByteArray()
    }

    /** 标准字母表编码，带 `=` 填充，与 `java.util.Base64` 的标准编码器一致。 */
    fun encode(value: ByteArray): String {
        if (value.isEmpty()) return ""
        val builder = StringBuilder((value.size + 2) / 3 * 4)
        var index = 0
        while (index + 2 < value.size) {
            val chunk = (value[index].toInt() and 0xFF shl 16) or
                (value[index + 1].toInt() and 0xFF shl 8) or
                (value[index + 2].toInt() and 0xFF)
            builder.append(ALPHABET[chunk shr 18 and 0x3F])
            builder.append(ALPHABET[chunk shr 12 and 0x3F])
            builder.append(ALPHABET[chunk shr 6 and 0x3F])
            builder.append(ALPHABET[chunk and 0x3F])
            index += 3
        }
        when (value.size - index) {
            1 -> {
                val chunk = value[index].toInt() and 0xFF shl 16
                builder.append(ALPHABET[chunk shr 18 and 0x3F])
                builder.append(ALPHABET[chunk shr 12 and 0x3F])
                builder.append("==")
            }

            2 -> {
                val chunk = (value[index].toInt() and 0xFF shl 16) or (value[index + 1].toInt() and 0xFF shl 8)
                builder.append(ALPHABET[chunk shr 18 and 0x3F])
                builder.append(ALPHABET[chunk shr 12 and 0x3F])
                builder.append(ALPHABET[chunk shr 6 and 0x3F])
                builder.append('=')
            }
        }
        return builder.toString()
    }
}
